go-controller/pkg/node/egressip.go

```diff
@@ -0,0 +1,186 @@
+// +build linux
+
+package node
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+	"time"
+
+	egressipv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1"
+	"github.com/vishvananda/netlink"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const egressLabel = "egress"
+
+func (n *OvnNode) WatchEgressIP() error {
+	_, err := n.watchFactory.AddEgressIPHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			eIP := obj.(*egressipv1.EgressIP)
+			if err := n.addEgressIP(eIP); err != nil {
+				klog.Error(err)
+			}
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldEIP := old.(*egressipv1.EgressIP)
+			newEIP := new.(*egressipv1.EgressIP)
+			if err := n.deleteEgressIP(oldEIP); err != nil {
+				klog.Error(err)
+			}
+			if err := n.addEgressIP(newEIP); err != nil {
+				klog.Error(err)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			eIP := obj.(*egressipv1.EgressIP)
+			if err := n.deleteEgressIP(eIP); err != nil {
+				klog.Error(err)
+			}
+		},
+	}, n.syncEgressIPs)
+	return err
+}
+
+func (n *OvnNode) isNodeIP(link netlink.Link, addr *netlink.Addr) (bool, error) {
+	nodeAddrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return true, err
+	}
+	for _, nodeAddr := range nodeAddrs {
+		if nodeAddr.Label != getEgressLabel(n.defaultGatewayIntf) {
+			if addr.IP.Equal(nodeAddr.IP) {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func (n *OvnNode) addHandler(link netlink.Link, egressIP string) error {
+	egressAddr, err := netlink.ParseAddr(fmt.Sprintf("%s/32", egressIP))
+	if err != nil {
+		return fmt.Errorf("unable to parse EgressIP: %s, err: %v", egressIP, err)
+	}
+	isNodeIP, err := n.isNodeIP(link, egressAddr)
+	if err != nil {
+		return fmt.Errorf("unable to list node's IPs")
+	}
+	if isNodeIP {
+		return fmt.Errorf("unable to add own node's IP to interface")
+	}
+	egressAddr.Label = n.defaultGatewayIntf + egressLabel
+	return netlink.AddrReplace(link, egressAddr)
```

Comment:
I think you need some more comments or some commit message for the "EgressIP node implementation". I'm not sure why we are watching from nodes, but it looks like so you can set the ip address on a link, which I don't think ovn-k8s should be doing. Especially not in shared GW mode.

Reply:
>  but it looks like so you can set the ip address on a link, which I don't think ovn-k8s should be doing

But attaching the egress IP address on the primary link is essentially the bare bones of this feature. How else would we the L3 SNAT work for south -> north traffic?

Reply:
So for shared gw mode, the IP should be on the GR. Then you should have SNAT configured on the GR to your IP for the pods affected by egress IP policy. There should be nothing done on the host itself.

For local gw mode, there's no real way around this you have to add the IP to the interface unfortunately.

Reply:
can you provide some comments on how you resolved this?

Reply:
It's turns out it's really easy. Each logical router is already aware of which network it belongs to, usually this is: `$NODE_IP/16`. All I thus needed to do was create a NAT rule without any additional network modification for this to work, as seen here: https://github.com/ovn-org/ovn-kubernetes/blob/353892570eb5218dfd26e7a783b4704330f35f4f/go-controller/pkg/ovn/egressip_shared.go#L61-L